youtube_dl/extractor/funimation.py

```diff
@@ -0,0 +1,110 @@
+# coding: utf-8
+from __future__ import unicode_literals
+from .common import InfoExtractor
+from ..utils import (sanitized_Request, ExtractorError, urlencode_postdata)
+
+import re
+
+class FunimationBaseIE(InfoExtractor):
+    _LOGIN_REQUIRED = True
+
+    def _login(self):
+        (username, password) = self._get_login_info()
+        if username is None:
+                if self._LOGIN_REQUIRED:
+                    raise ExtractorError('No login info available, needed for using %s.' % self.IE_NAME, expected=True)
+                return
+        self.report_login()
+        login_url = 'http://www.funimation.com/login'
+        data = urlencode_postdata({
+            'formname': 'loginForm2',
+            'email_field': username,
+            'password_field': password,
+        })
+        login_request = sanitized_Request(login_url, data)
+        login_request.add_header('Content-Type', 'application/x-www-form-urlencoded')
+        self._download_webpage(login_request, login_url)
+
+    def _real_initialize(self):
+        self._login()
+
+class FunimationIE(FunimationBaseIE):
+    _VALID_URL = r'(?P<address>https?://(?:www\.)?funimation\.com/shows/.*/videos/official/.+)'
+    _USER_AGENT = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/601.2.7 (KHTML, like Gecko) Version/9.0.1 Safari/601.2.7'
+
+    _TEST = {
+        'url': 'http://www.funimation.com/shows/noragami/videos/official/a-housecat-a-stray-god-and-a-tail?watch=sub',
+        'info_dict': {
+            'id': 'NRGENG0001',
+            'title': 'Noragami - A Housecat, a Stray God, and a Tail',
+            'ext': 'mp4',
+        },
+        'params': {
+            'username': 'foo@bar.com',
+            'password': '(snip)',
+        },
+        'skip': 'requires actual password',
+    }
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+
+        request = sanitized_Request(url)
+        request.add_header('User-Agent', self._USER_AGENT)
+
+        video_address = mobj.group('address')
+        webpage = self._download_webpage(request, video_address)
```

Comment:
I tried to use taht first, but couldn't handle the regex. The url sometimes end with title-of-ep, and sometimes with title-of-ep?watch=sub and title-of-ep?watch=dub

I wasn't able to make the last part of the regex optional.